### PR TITLE
[SPARK-30848][SQL] Remove productHash from TreeNode

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to replace `TreeNode.productHash` with `MurmurHash3.productHash`.

### Why are the changes needed?
1.Spark 4.0 no longer supports Scala 2.12.

2.Using MurmurHash3's class method `productHash` to reduce code redundancy.
https://github.com/scala/scala/blob/v2.13.11/src/library/scala/util/hashing/MurmurHash3.scala#L343
https://github.com/scala/scala/blob/v2.13.11/src/library/scala/util/hashing/MurmurHash3.scala#L64-L81

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
1.Pass GA.
2.Manually test:
```
(base) panbingkun:~/Developer/spark/spark-community$./build/sbt "catalyst/testOnly org.apache.spark.sql.catalyst.expressions.CanonicalizeSuite -- -t \"SPARK-30847: Take productPrefix into account in MurmurHash3.productHash\""

[info] CanonicalizeSuite:
[info] - SPARK-30847: Take productPrefix into account in MurmurHash3.productHash (249 milliseconds)
[info] Run completed in 1 second, 135 milliseconds.
[info] Total number of tests run: 1
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 1, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 230 s (03:50), completed Sep 28, 2023, 12:54:06 PM
```

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by LuciferYang** [comment]
@panbingkun Could you resolve the conflict?

**Comment by panbingkun** [comment]
> @panbingkun Could you resolve the conflict?

@LuciferYang Done.

**Review by beliefer — COMMENTED** [review]
Looks good.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala [hunk 1]
```diff
@@ -175,30 +176,7 @@ abstract class TreeNode[BaseType <: TreeNode[BaseType]]
 
   lazy val height: Int = children.map(_.height).reduceOption(_ max _).getOrElse(0) + 1
 
-  // Copied from Scala 2.13.1
-  // github.com/scala/scala/blob/v2.13.1/src/library/scala/util/hashing/MurmurHash3.scala#L56-L73
-  // to prevent the issue https://github.com/scala/bug/issues/10495
-  // TODO(SPARK-30848): Remove this once we drop Scala 2.12.
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:181]: You can use https://issues.apache.org/jira/browse/SPARK-30848 JIRA :-).

**Review by HyukjinKwon — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala (hunk 1 above)
Comment (panbingkun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/trees/TreeNode.scala:181]: Yeah, I was looking for this PR to learn about history. Thank you for the good record at that time! haha

**Review by beliefer — APPROVED** [review]

**Review by LuciferYang — APPROVED** [review]
+1, LGTM

**Pull request closed without merging**